File: sites/bestBuy.js
const delay = require('delay');
const cheerio = require('cheerio');
const axios = require('axios').default;
const fs = require('fs');
const { gzip } = require('zlib');
const { response } = require('express');
const { resolve } = require('path');
const { json } = require('body-parser');
const fetch = require('node-fetch');
const Discord = require('discord.js');
const got = require('got').default
const HttpsProxyAgent = require('https-proxy-agent')

// require ('newrelic');
// const rp = require('request-promise').defaults({
// 	// followAllRedirects: true,
// 	resolveWithFullResponse: true,
// 	gzip : true,
//     timeout: 5000
// });
// Production https://discordapp.com/api/webhooks/827263591114997790/chAZK84Gnad7rjHDlh4BnF7dz5KQ7-0l4atsFzJGgcTkAaeZno6ePYB_A-WiiClS3FpY
 const webhookClient1 = new Discord.WebhookClient('827263591114997790', 'chAZK84Gnad7rjHDlh4BnF7dz5KQ7-0l4atsFzJGgcTkAaeZno6ePYB_A-WiiClS3FpY');

// Test https://discordapp.com/api/webhooks/745279081247014942/3TuT8vs6BUXr9HAK1uRKaB4t3Ap0LnoLfPJTgT1uhNzQvqR1GsUXW-d4_dxCrgOCdkBM

class bestBuyMonitor {
    constructor(sku) {
        this.trueSku = sku
        this.sku = sku.split(':')[0];
        this.skuName = sku.split(':')[1]
        this.delay = 850000; // this.delay = 390000
        this.startDelay = 800; //  this.startDelay = 6000;
        this.availability = '';
        this.stockNumber = '';
        this.proxyList = [];
        this.isStock = false
        this.imageUrl = ''
        this.productName = ''

    }

    async task () {
        try {
            console.log('Start')
            console.log(this.sku)
            console.log(this.skuName)
            await this.getProxies()
         //   console.log(this.proxyList)
            await this.monitor()
        } catch (error) {
            fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                console.log(err)
            }))
        }
    }
    async getProxies() {
		try {
			// read contents of the file
                const data = fs.readFileSync('bestBuyProxies.txt', 'utf-8');

			const lines = data.split(/\r?\n/);

			// print all lines
			lines.forEach((line) => {
				const lineSplit = line.split(':');
				const item1 = {
					ip : lineSplit[0],
					port: lineSplit[1],
					userAuth: lineSplit[2],
					userPass: lineSplit[3],
				};
				this.proxyList.push(item1);

				// console.log(line);
				// console.log(item1);
				// console.log('\n\n\n\n\n');
			});
		} catch (err) {
			console.error(err);
            fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                console.log(err)
            }))
        }
    }
    async getPic(){
                try {
                   var axios = require("axios").default;

                var options = {
                method: 'GET',
                url: `https://www.bestbuy.com/site/prada/${this.sku}.p`,
                params: {skuId: `${this.sku}`},
                };

                let fetchSite = await axios.request(options).then(response => {
                   return response?.data?.split('","image":"')[1]?.split('"')[0]
                }).catch(err => {
                    console.log(err)
                })
                console.log(fetchSite)
                this.image = fetchSite
                
                
                console.log(this.image)
                } catch (error) {
                    console.log(`Error on getting pic ${error.message}`)
                    this.getPic
                }
    }
    async monitor () {
       try {
        console.log('Starting Monitoring')
        var testing = ''
        return new Promise( async ( resolve, reject) => {

                let i = 0
                var { skuBank } = require('../dms')
                let index = skuBank.findIndex(e => e.sku == this.trueSku)
                    this.getPic()
             
                    while(!skuBank[index]?.stop){
                            let timeStart = Date.now();
                            let timeStartSec = (timeStart / 1000);
                    skuBank[index].name = this.productName
                        if(i+1 == this.proxyList.length){
                            i = 0
                        }
                        let proxy = this.proxyList[i]
                        i++
                      //  console.log(`${proxy.userAuth}:${proxy.userPass}@${proxy.ip}:${proxy.port}`);
                        (async () =>{
                              try {
                               var options = {
                                            method: 'GET',
                                            url: 'https://www.bestbuy.com/api/3.0/priceBlocks',
                                            params: {skus: `${this.sku}`},
                                                       headers: {
                                                                "accept": "*/*",
                                                                        "accept-language": "en-US,en;q=0.9",
                                                                        "cache-control": "no-cache",
                                                                        "pragma": "no-cache",
                                                                        "sec-fetch-dest": "empty",
                                                                        "sec-fetch-mode": "cors",
                                                                        "sec-fetch-site": "same-origin",
                                                                        "user-agent" : "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
                                                                },
                                             proxy: {
                                                        host: proxy.ip,
                                                        port: proxy.port,
                                                        auth: {
                                                        username: proxy.userAuth,
                                                        password: proxy.userPass
                                                        }
                                                    },
                                            timeout : 1000
                                            };
                                            

                                let fetchSite = await axios.request(options)
                                // console.log(fetchSite?.data)
                                let parsed = fetchSite?.data
                                let availability = parsed[0]?.sku?.buttonState?.buttonState
                                this.productName = parsed[0]?.sku?.names?.short
                                let price = parsed[0]?.sku?.price?.currentPrice
                                // let availability = fetchSite?.body?.split(',"availability":"')[1]?.split('.org/')[1]?.split('"')[0]
                                // this.productName = fetchSite?.body?.split('"@context":"http://schema.org/","@type":"Product","name":"')[1]?.split('"')[0]
                                // let price = fetchSite?.body?.split('"@context":"http://schema.org/","@type":"Product","name":"')[1]?.split('price":"')[1]?.split('"')[0]
                                // 
                                if(availability == "SOLD_OUT"){
                                    this.availability = false
                                } else if (availability == "ADD_TO_CART"){
                                    this.availability = true
                                } else {
                                    console.log(availability)
                                }
                                
                                console.log(`Task ${i} : ${fetchSite.status}`, this.availability, this.productName, price , this.isStock)
                                if(!this.isStock && this.availability) {
                                    // Send in stock webhook
                                    console.log(`Task ${i} : ${this.isStock} and ${this.availability}`)
                                    this.isStock = true
                                    let embed1 = new Discord.MessageEmbed()
                                    .setColor('#07bf6e')
                                    .setTitle('Best Buy Monitor')
                                    .setURL(`https://www.bestbuy.com/site/prada/${this.sku}.p?skuId=${this.sku}`)
                                    .addFields(
                                        { name : 'Product Name', value : `${this.productName}`},
                                        { name : 'Product Availability', value : `Product In Stock`, inline : true},
                                    //    { name : 'Stock Number', value : `${this.stockNumber}`, inline : true}, 
                                        { name : 'Current Price', value : price, inline : true} 
                                        )
                                    .addField("Links", `[ATC](https://api.bestbuy.com/click/tempo/${this.sku}/cart) | [Cart](https://www.bestbuy.com/cart)`)
                                    // .addField('Original Price', originalPrice, true)
                                    // .addField('Current Price', currentPrice, true)
                                    .setThumbnail(this.image)
                                    .setTimestamp()
                                    .setFooter('Prada#4873', 'https://cdn.discordapp.com/attachments/772173046235529256/795132477659152444/pradakicks.jpg');
                                    webhookClient1.send({
                                        username: 'Best Buy',
                                        avatarURL: 'https://cdn.discordapp.com/attachments/815507198394105867/816741454922776576/pfp.png',
                                        embeds: [embed1],
                                    })
                                } else if (!this.availability && this.isStock) {
                                    this.isStock = false
                                } else if (!this.availability && !this.isStock){
                                    this.isStock = false
                                    this.availability = false
                                    // Not Important
                                  //  console.log('False false')
                                } else if(this.availability && this.isStock){
                                    this.availability = true
                                    this.isStock = true
                                    // Not Important
                                    // console.log(true, true)
                                } else {
                                    fs.appendFileSync('what.txt', this.availability + this.isStock + '\n', (err =>{
                                        console.log(err)
                                    }))
                                }
            
                            
                                // console.log(originalPrice, currentPrice)
                               // console.log(parsedBod)
                
                
                            } catch (error) {
                                console.log(error)
                                // fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                                //     console.log(err)
                                // }))
                                
                                if(error.message.includes('Unexpected token')){
                                      if(testing.includes('<!DOCTYPE html><html')){
                                        console.log("HTML")
                                      } else if (testing.includes('Are you a human?')) {
                                          console.log('Captcha')
                                      } else {
                                            console.log('No HTML')
                                            console.log(testing)
                                      }
                                   
                                } else if (error.message.includes('403')){
                                    // await delay(400000)
                                    console.log('403 Access Denied')
                                }
                            }
                        })()
                        

                          
                     
                        // let timeEnd = Date.now();
                        // let timeEndSec = timeEnd / 1000
                        // console.log(timeEndSec - timeStartSec)
                        await delay(this.startDelay)
                    }
                    console.log('stopped!')
                    resolve('Stopped')
                    return
              
        })
       } catch (error) {
        fs.appendFileSync('bigError.txt', error.toString() + '\n', (err =>{
            console.log(err)
        }))
       }
  
    }

}

// const monitoring = new newEggMonitor(`19-113-569`);

// (async ()=>{
//     await monitoring.task()
// })()

module.exports = {
    bestBuyMonitor
}

File: GoMonitor/sites/bestBuy/bestBuy.go
package BestBuyMonitor

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/bradhe/stopwatch"
	FetchProxies "github.con/prada-monitors-go/helpers/proxy"
	Types "github.con/prada-monitors-go/helpers/types"
)

type Config struct {
	sku              string
	skuName          string // Only for new Egg
	startDelay       int
	discord          string
	site             string
	priceRangeMax    int
	priceRangeMin    int
	proxyCount       int
	indexMonitorJson int
}
type Monitor struct {
	Config              Config
	monitorProduct      Product
	Availability        string
	currentAvailability string
	Client              http.Client
	file                *os.File
	stop                bool
	CurrentCompanies    []Company
}
type Product struct {
	name        string
	stockNumber string
	productId   string
	price       int
	image       string
	link        string
}
type Proxy struct {
	ip       string
	port     string
	userAuth string
	userPass string
}
type ItemInMonitorJson struct {
	Sku       string `json:"sku"`
	Site      string `json:"site"`
	Stop      bool   `json:"stop"`
	Name      string `json:"name"`
	Companies []Company
}
type Company struct {
	Company      string `json:"company"`
	Webhook      string `json:"webhook"`
	Color        string `json:"color"`
	CompanyImage string `json:"companyImage"`
}
type bestBuyResponse []struct {
	Sku struct {
		ButtonState struct {
			ButtonState string `json:"buttonState"`
			DisplayText string `json:"displayText"`
			SkuID       string `json:"skuId"`
		} `json:"buttonState"`
		Condition        string        `json:"condition"`
		InkSubscriptions []interface{} `json:"inkSubscriptions"`
		Names            struct {
			Short string `json:"short"`
		} `json:"names"`
		Price struct {
			CurrentPrice float64 `json:"currentPrice"`
			PriceDomain  struct {
				CurrentAsOfDate string  `json:"currentAsOfDate"`
				CurrentPrice    float64 `json:"currentPrice"`
				IsMAP           bool    `json:"isMAP"`
				PriceEventType  string  `json:"priceEventType"`
				RegularPrice    float64 `json:"regularPrice"`
				SkuID           string  `json:"skuId"`
			} `json:"priceDomain"`
			PricingType        string `json:"pricingType"`
			SmartPricerEnabled bool   `json:"smartPricerEnabled"`
		} `json:"price"`
		ProductType string `json:"productType"`
		SkuID       string `json:"skuId"`
		Subclass    struct {
			DisplayName string `json:"displayName"`
			ID          string `json:"id"`
		} `json:"subclass"`
		URL string `json:"url"`
	} `json:"sku"`
}

func NewMonitor(sku string) *Monitor {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", sku, sku, r)
		}
	}()
	fmt.Println("TESTING")
	m := Monitor{}
	m.Availability = "SOLD_OUT"
	// var err error
	//	m.Client = http.Client{Timeout: 10 * time.Second}
	m.Config.site = "BestBuy"
	m.Config.startDelay = 3000
	m.Config.sku = sku
	// 	m.file, err = os.Create("./testing.txt")
	m.Client = http.Client{Timeout: 10 * time.Second}
	m.Config.discord = "https://discord.com/api/webhooks/827263591114997790/chAZK84Gnad7rjHDlh4BnF7dz5KQ7-0l4atsFzJGgcTkAaeZno6ePYB_A-WiiClS3FpY"
	m.monitorProduct.name = "Testing Product"
	m.monitorProduct.stockNumber = ""
	m.getProductDetails()

	proxyList := FetchProxies.Get()

	// fmt.Println(timeout)
	//m.Availability = "OUT_OF_STOCK"
	//fmt.Println(m)
	i := true
	time.Sleep(15000 * (time.Millisecond))
	go m.checkStop()
	time.Sleep(3000 * (time.Millisecond))
	for i {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
			}
		}()

		if !m.stop {
			currentProxy := m.getProxy(proxyList)
			splittedProxy := strings.Split(currentProxy, ":")
			proxy := Proxy{splittedProxy[0], splittedProxy[1], splittedProxy[2], splittedProxy[3]}
			prox1y := fmt.Sprintf("http://%s:%s@%s:%s", proxy.userAuth, proxy.userPass, proxy.ip, proxy.port)
			proxyUrl, err := url.Parse(prox1y)
			if err != nil {
				fmt.Println(err)
				return nil
			}
			defaultTransport := &http.Transport{
				Proxy: http.ProxyURL(proxyUrl),
			}
			m.Client.Transport = defaultTransport
			m.monitor()
			// time.Sleep(150 * (time.Millisecond))
		} else {
			fmt.Println(m.Config.sku, "STOPPED STOPPED STOPPED")
			i = false
		}

	}
	return &m
}
func (m *Monitor) monitor() error {
	watch := stopwatch.Start()
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
		}
	}()

	url := fmt.Sprintf("https://www.bestbuy.com/api/3.0/priceBlocks?skus=%s", m.Config.sku)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	req.Header.Add("pragma", "no-cache")
	req.Header.Add("cache-control", "no-cache")
	req.Header.Add("accept", "application/json")
	req.Header.Add("dnt", "1")
	req.Header.Add("accept-language", "en")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36")
	req.Header.Add("content-type", "application/json")
	req.Header.Add("sec-fetch-site", "cross-site")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-dest", "empty")
	req.Header.Set("Connection", "close")
	req.Close = true
	res, err := m.Client.Do(req)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	defer res.Body.Close()
	defer func() {
		watch.Stop()
		fmt.Printf("Best Buy - Status Code : %d  %s, %s : : Milliseconds elapsed: %v \n", res.StatusCode, m.Availability, m.Config.sku, watch.Milliseconds())
	}()
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	if res.StatusCode != 200 {
		return nil
	}
	var realBody bestBuyResponse
	err = json.Unmarshal([]byte(body), &realBody)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	var monitorAvailability string
	monitorAvailability = realBody[0].Sku.ButtonState.ButtonState
	m.monitorProduct.name = realBody[0].Sku.Names.Short
	m.monitorProduct.price = int(realBody[0].Sku.Price.CurrentPrice)
	if monitorAvailability != "ADD_TO_CART" {
		monitorAvailability = "SOLD_OUT"
	}

	if m.Availability == "SOLD_OUT" && monitorAvailability == "ADD_TO_CART" {
		fmt.Println("Item in Stock")
		go m.sendWebhook()
	}
	if m.Availability == "ADD_TO_CART" && monitorAvailability == "SOLD_OUT" {
		fmt.Println("Item Out Of Stock")
	}
	m.Availability = monitorAvailability
	return nil
}
func (m *Monitor) getProxy(proxyList []string) string {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
		}
	}()
	//fmt.Scanln()
	// rand.Seed(time.Now().UnixNano())
	// randomPosition := rand.Intn(len(proxyList)-0) + 0
	if m.Config.proxyCount+1 == len(proxyList) {
		m.Config.proxyCount = 0
	}
	m.Config.proxyCount++
	//fmt.Println(proxyList[m.Config.proxyCount])
	return proxyList[m.Config.proxyCount]
}
func (m *Monitor) sendWebhook() error {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
		}
	}()
	// now := time.Now()
	// currentTime := strings.Split(now.String(), "-0400")[0]
	for _, letter := range m.monitorProduct.name {
		if string(letter) == `"` {
			m.monitorProduct.name = strings.Replace(m.monitorProduct.name, `"`, "", -1)
		}
	}
	fmt.Println("Testing Here : ", m.monitorProduct.name, "Here")
	if strings.HasSuffix(m.monitorProduct.name, "                       ") {
		m.monitorProduct.name = strings.Replace(m.monitorProduct.name, "                       ", "", -1)
	}
	t := time.Now().UTC().Format("2006-01-02T15:04:05Z")

	fmt.Println("Testing Here : ", m.monitorProduct.name, "Here")
	for _, comp := range m.CurrentCompanies {
		fmt.Println(comp.Company)
		go webHookSend(comp, m.Config.site, m.Config.sku, m.monitorProduct.name, m.monitorProduct.price, t, m.monitorProduct.image)
	}
	// payload := strings.NewReader("{\"content\":null,\"embeds\":[{\"title\":\"Target Monitor\",\"url\":\"https://discord.com/developers/docs/resources/channel#create-message\",\"color\":507758,\"fields\":[{\"name\":\"Product Name\",\"value\":\"%s\"},{\"name\":\"Product Availability\",\"value\":\"In Stock\\u0021\",\"inline\":true},{\"name\":\"Stock Number\",\"value\":\"%s\",\"inline\":true},{\"name\":\"Links\",\"value\":\"[Product](https://www.walmart.com/ip/prada/%s)\"}],\"footer\":{\"text\":\"Prada#4873\"},\"timestamp\":\"2021-04-01T18:40:00.000Z\",\"thumbnail\":{\"url\":\"https://cdn.discordapp.com/attachments/815507198394105867/816741454922776576/pfp.png\"}}],\"avatar_url\":\"https://cdn.discordapp.com/attachments/815507198394105867/816741454922776576/pfp.png\"}")
	return nil
}
func (m *Monitor) getProductDetails() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
		}
	}()

	url := fmt.Sprintf("https://www.bestbuy.com/site/prada/%s.p", m.Config.sku)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Println(err)

		return
	}
	req.Header.Add("pragma", "no-cache")
	req.Header.Add("cache-control", "no-cache")
	req.Header.Add("accept", "application/json, text/plain, */*")
	req.Header.Add("dnt", "1")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36")
	req.Header.Add("sec-fetch-site", "same-origin")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-dest", "empty")
	req.Header.Add("accept-language", "en-US,en;q=0.9")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Println(err)

		return
	}
	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		log.Fatal(err)
	}
	res.Body.Close()
	m.monitorProduct.image = doc.Find(".primary-image").AttrOr("src", "https://cdn.discordapp.com/attachments/815507198394105867/816741454922776576/pfp.png")
	fmt.Println(m.monitorProduct.image)
}
func webHookSend(c Company, site string, sku string, name string, price int, time string, image string) {
	payload := strings.NewReader(fmt.Sprintf(`{
		"content": null,
		"embeds": [
		  {
			"title": "%s Monitor",
			"url": "https://www.bestbuy.com/site/prada/%s.p?skuId=%s",
			"color": %s,
			"fields": [
			  {
				"name": "Product Name",
				"value": "%s"
			  },
			  {
				"name": "Product Availability",
				"value": "In Stock",
				"inline": true
			  },
			  {
				"name": "Price",
				"value": "%d",
				"inline": true
			  },
			  {
				"name": "Product Sku",
				"value": "%s",
				"inline": true
			  },
			  {
				"name": "Links",
				"value": "[ATC](https://api.bestbuy.com/click/tempo/%s/cart) | [Cart](https://www.bestbuy.com/cart)"
			  }
			],
			"footer": {
			  "text": "Prada#4873"
			},
			"timestamp": "%s",
			"thumbnail": {
			  "url": "%s"
			}
		  }
		],
		"avatar_url": "%s"
	  }`, site, sku, sku, c.Color, name, price, sku, sku, time, image, c.CompanyImage))
	req, err := http.NewRequest("POST", c.Webhook, payload)
	if err != nil {
		fmt.Println(err)
	}
	req.Header.Add("pragma", "no-cache")
	req.Header.Add("cache-control", "no-cache")
	req.Header.Add("accept", "application/json")
	req.Header.Add("dnt", "1")
	req.Header.Add("accept-language", "en")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36")
	req.Header.Add("content-type", "application/json")
	req.Header.Add("sec-fetch-site", "cross-site")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-dest", "empty")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Println(err)
	}
	defer res.Body.Close()
	fmt.Println(res)
	fmt.Println(payload)
	return
}
func (m *Monitor) checkStop() error {
	for !m.stop {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Site : %s, Product : %s  CHECK STOP Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
			}
		}()
		getDBPayload := strings.NewReader(fmt.Sprintf(`{
			"site" : "%s",
			"sku" : "%s"
		  }`, strings.ToUpper(m.Config.site), m.Config.sku))
		url := "http://localhost:7243/DB"
		req, err := http.NewRequest("POST", url, getDBPayload)
		if err != nil {
			fmt.Println(err)
			return nil
		}
		res, err := http.DefaultClient.Do(req)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil

		}
		body, err := ioutil.ReadAll(res.Body)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil
		}
		var currentObject ItemInMonitorJson
		err = json.Unmarshal(body, &currentObject)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil
		}
		m.stop = currentObject.Stop
		m.CurrentCompanies = currentObject.Companies
		fmt.Println(m.CurrentCompanies)
		res.Body.Close()
		time.Sleep(5000 * (time.Millisecond))
	}
	return nil
}
func BestBuy(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	fmt.Fprintf(w, "Best Buy Monitor")
	fmt.Println("Best Buy")
	var currentMonitor Types.MonitorResponse
	_ = json.NewDecoder(r.Body).Decode(&currentMonitor)
	fmt.Println(currentMonitor)
	go NewMonitor(currentMonitor.Sku)
	json.NewEncoder(w).Encode(currentMonitor)
}
